home-manager/dotfiles/doom.d/config.org
#+title: Emacs Literate Config
#+author: Alan Hettinger

** Basic settings
*** Personal information
#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

(setq user-full-name "Alan Hettinger"
      user-mail-address "alan.hettinger@proton.me")
#+end_src
*** Fonts
*Note*: ~doom-font~ and so on treat an integer ~:size~ as pixel size and float as point size.
Fonts I've liked:
 * Fragment Mono
 * Jetbrains Mono
 * Variable: et-book
 * mononoki
Options that are not currently set:
 * =(setq doom-big-font (font-spec :family "" :size 20))=
 * =(setq doom-serif font "")=
 #+begin_src emacs-lisp
(setq doom-font (font-spec :family "mononoki" :size 16.0))
(setq doom-variable-pitch-font (font-spec :family "Source Serif Variable" :height 1.2 :size 16.0))

(use-package! mixed-pitch
  :config
  (setq mixed-pitch-set-height t)
  (set-face-attribute 'variable-pitch nil :height 1.2)
  )
 #+end_src
*** Theme and global decor
#+begin_src emacs-lisp
;; (load-theme 'catppuccin t t)
;; (setq catppuccin-flavor 'macchiato)
(use-package! catppuccin-theme
  :config (load-theme 'catppuccin t)
  :init (setq catppuccin-flavor 'macchiato))
;; (catppuccin-reload)

(setq doom-theme 'catppuccin)
;; (setq doom-themes-padded-modeline t)

;; (setq fancy-splash-image "~/.doom.d/Emacs-logo.png")

(setq display-line-numbers-type 'relative)

;; disable title bar in gnome
(setq default-frame-alist '((undecorated . t)))
(setq lsp-headerline-breadcrumb-enable t)
#+end_src
**** Mode line
#+begin_src emacs-lisp
(after! doom-modeline
  (setq doom-modeline-enable-word-count t
        doom-modeline-major-mode-icon t
        doom-modeline-persp-name t
        doom-modeline-height 35
        doom-modeline-icon t
        doom-modeline-modal-modern-icon nil
        doom-modeline-buffer-modification-icon t
        doom-modeline-highlight-modified-buffer-name nil
        )
  (remove-hook 'doom-modeline-mode-hook #'size-indication-mode)
  (remove-hook 'doom-modeline-mode-hook #'column-number-mode)
  (line-number-mode -1)
)
#+end_src
**** Make the ascii banner look nicer:
credit to https://tecosaur.github.io/emacs-config/config.html#ascii-banner
#+begin_src emacs-lisp
(defun doom-custom-banner ()
  (let* ((banner
          '(",---.,-.-.,---.,---.,---."
            "|---'| | |,---||    `---."
            "`---'` ' '`---^`---'`---'"))
         (longest-line (apply #'max (mapcar #'length banner))))
    (put-text-property
     (point)
     (dolist (line banner (point))
       (insert (+doom-dashboard--center
                +doom-dashboard--width
                (concat line (make-string (max 0 (- longest-line (length line))) 32))) "\n"))
       'face 'doom-dashboard-banner)))

(setq +doom-dashboard-ascii-banner-fn #'doom-custom-banner)
#+end_src
**** Toggle light/dark mode on the fly
#+begin_src emacs-lisp
(defun toggle-catppuccin-light-dark-theme ()
  "toggle between light and dark catppuccin colorschemes"
  (interactive (progn
                  (let ((active-theme catppuccin-flavor)
                        (light-theme 'latte)
                        (dark-theme 'macchiato))
                    ((lambda (desired-theme) (and (setq catppuccin-flavor desired-theme) (catppuccin-reload)))
                     (if (eq active-theme dark-theme) light-theme dark-theme)))
                  nil)))

(defun toggle-light-dark-theme ()
  "toggle between light and dark theme generally"
  (interactive (progn
                 (let* ((active-theme doom-theme)
                       (light-theme 'doom-gruvbox-light)
                       (dark-theme 'catppuccin))
                   (load-theme (if (eq doom-theme dark-theme) light-theme dark-theme)))
                 nil)))

(map! :leader
      (:prefix "t"
               :desc "toggle light/dark mode" "d" #'toggle-light-dark-theme))
#+end_src
**** Frame title format
Remove " - Doom Emacs" from the frame title (the window name, in normal terms)
#+begin_src emacs-lisp
(setq frame-title-format '("%b"))
#+end_src
*** disable line numbers in some modes
#+begin_src emacs-lisp
(dolist (mode '(org-mode-hook
                vterm-mode-hook
                eshell-mode-hook
                markdown-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))
#+end_src
*** no scroll bars ever
#+begin_src emacs-lisp
(defun my/disable-scroll-bars (frame)
  (modify-frame-parameters frame
                           '((vertical-scroll-bars . nil)
                             (horizontal-scroll-bars . nil))))
(add-hook 'after-make-frame-functions 'my/disable-scroll-bars)
#+end_src
*** Better visual lines:
#+begin_src emacs-lisp
(use-package! evil-better-visual-line
  :config (evil-better-visual-line-on))
(+global-word-wrap-mode +1)
#+end_src
*** Splits
#+begin_src emacs-lisp
(setq split-height-threshold nil
       split-width-threshold 40)
#+end_src
*** Misc global settings
#+begin_src emacs-lisp
(setq company-minimum-prefix-length 3)

(global-auto-revert-mode 1)
(setq global-auto-revert-non-file-buffers t)

(setq ispell-dictionary "en_US")
(setq +word-wrap-disabled-modes '(vterm-mode))
#+end_src
#+begin_src emacs-lisp
(setq evil-want-fine-undo t
      )
#+end_src

**** TODO break these down under headers
*** Keybindings
Show/hide modeline:
#+begin_src emacs-lisp
(map! :leader
      (:prefix "t"
               :desc "toggle modeline"  "m" #'hide-mode-line-mode)
      (:prefix "q"
               :desc "save and quit server-edit frame" "e" #'server-edit
               :desc "abort server-edit frame" "E" #'server-edit-abort)
                )
#+end_src
Bind alt-tab to company completion rather than spell-checking:
#+begin_src emacs-lisp
(map! :i "M-TAB" (cmds! (not (minibufferp)) #'company-complete-common))
#+end_src
*** TODO Perspectives and tab bar
show workspace names in tab bar. Per https://github.com/Bad-ptr/persp-mode.el/issues/122#issuecomment-1224884651. Not currently working on my setup.
#+begin_src emacs-lisp :tangle no
  (add-hook 'persp-before-deactivate-functions
            (defun +workspaces-save-tab-bar-data-h (_)
              (when (get-current-persp)
                (set-persp-parameter
                 'tab-bar-tabs (tab-bar-tabs)))))

  (add-hook 'persp-activated-functions
            (defun +workspaces-load-tab-bar-data-h (_)
              (tab-bar-tabs-set (persp-parameter 'tab-bar-tabs))
              (tab-bar--update-tab-bar-lines t)))
#+end_src
*** PDF mode
#+begin_src emacs-lisp
(add-hook 'pdf-view-mode-hook
          (lambda () (pdf-view-auto-slice-minor-mode 1)))
#+end_src
** Prose (org and markdown)
*** General settings
#+begin_src emacs-lisp
;; (setq company-global-modes '(not org-mode, not markdown-mode not eshell-mode))

(dolist (mode '(org-mode-hook
                markdown-mode-hook
                Info-mode-hook))
        (add-hook mode (lambda () (mixed-pitch-mode 1) ))
        (add-hook mode (lambda () (progn
                         (setq left-margin-width 4)
                         (setq right-margin-width 4)
                         (set-window-buffer nil (current-buffer))))))

(after! org
  (setq org-startup-folded t)
  (setq org-directory "~/Documents/Notes/org/")
  ;; makes info files linkable from org
   (add-to-list 'org-modules 'ol-info)

  (add-hook 'org-mode-hook
	    (setq org-pretty-entities t
		  org-hide-emphasis-markers t
		  org-adapt-indentation t))
  (setq org-ellipsis " ▼ ")
                )
#+end_src
*** TODO Company-ispell
For some reason this doesn't do what it is supposed to
#+begin_src emacs-lisp
(add-hook 'org-mode-hook
          (lambda () (setq-local company-idle-delay nil)))
#+end_src
*** Export settings
#+begin_src emacs-lisp
(after! org
(setq org-export-with-section-numbers nil
      org-export-with-toc nil
      ;; org-odt-preferred-output-format docx
      )
                )
#+end_src
The following allows headlines to be ignored while including their contents:
#+begin_src emacs-lisp
(defun org-remove-headlines (backend)
  (org-map-entries (lambda () (delete-region (point-at-bol) (point-at-eol)))
                   "ignore"))

(add-hook 'org-export-before-processing-hook #'org-remove-headlines)
#+end_src
Automatically export my thesis draft as a pdf:
#+begin_src emacs-lisp
(after! org
  (let* ((main-filename-str "thesis-draft")
         (date-format-str (format-time-string "%Y-%m-%d"))
         (target-path "./drafts/")
         (target-filename (format "%s%s-%s.tex" target-path date-format-str main-filename-str))
        )
   (defun alan/export-thesis-draft ()
    (when (string-equal (buffer-file-name) (expand-file-name "./thesis-draft.org"))
      (progn
      ;; create the pdf:
      (org-export-to-file 'latex
          (format "./drafts/%s-thesis-draft.tex" (format-time-string "%Y-%m-%d"))
        nil ;; export asynchronously
        nil nil nil nil ;; bunch of args I don't care about
        #'org-latex-compile)
      ;; clean up the latex file:
      (delete-file (format "./drafts/%s-thesis-draft.tex" (format-time-string "%Y-%m-%d"))))
      )
    ))
  (add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook #'alan/export-thesis-draft)))
                )
#+end_src

*** Org-babel
#+begin_src emacs-lisp
(org-babel-do-load-languages
 'org-babel-load-languages
 '((emacs-lisp . t)
   (lua . t)
   (scheme . t)
   (shell . t)
   (racket . t)
   ))

(after! org
  (require 'org-tempo)
  (add-to-list 'org-structure-template-alist '("sh" . "src shell"))
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("lua" . "src lua"))
  (add-to-list 'org-structure-template-alist '("sc" . "src scheme"))
  (add-to-list 'org-structure-template-alist '("r" . "src racket"))
  ;; automatically tangle certain config files on save:
  ;; (defun alan/org-babel-tangle-config ()
  ;;   (when (string-equal (buffer-file-name)
  ;;                       (expand-file-name "./config.org"))
  ;;     (let ((org-confirm-babel-evaluate nil))
  ;;       (org-babel-tangle))))
  ;; (add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook #'alan/org-babel-tangle-config)))
)
#+end_src
Racket configuration for SICP:
#+begin_src emacs-lisp
(use-package! ob-racket
  :after org
;;   :config (add-hook 'ob-racket-pre-runtime-library-load-hook
;;               #'ob-racket-raco-make-runtime-library)
  )
;; (setq ob-racket-default-lang "sicp")
#+end_src
*** Citations
#+begin_src emacs-lisp
(after! org

  (setq org-latex-packages-alist '(("margin=2cm" "geometry" nil)))
  (setq org-cite-global-bibliography '("~/Documents/Thesis/zotero-lib.bib")
	org-cite-insert-processor 'citar
	org-cite-follow-processor 'citar
        org-cite-activate-processor 'citar
        citar-bibliography org-cite-global-bibliography
        org-cite-csl-styles-dir "~/Zotero/styles/"
        citar-citeproc-csl-styles-dir "~/Zotero/styles/"
        org-cite-export-processors
        '((latex . (csl "chicago-author-date.csl"))
          (odt . (csl "chicago-author-date.csl"))
          (t . (csl "chicago-author-date.csl")))
        )
)
#+end_src
*** Zen mode
Unused settings:
 * =(setq writeroom-width 0.4)=
 * =(setq writeroom-extra-line-spacing 0.2)=
 * =(setq writeroom-mode-line 't)=
 * =(setq writeroom-bottom-divider-width 0)=
#+begin_src emacs-lisp
(after! writeroom-mode
  (setq +zen-text-scale 1)
  (setq writeroom-mode-line 't)
  )
#+end_src
Tell zen mode to "disable" mixed-pitch-mode in org, because it is already an org hook:
#+begin_src emacs-lisp
;; (setq +zen-mixed-pitch-modes 'nil)
#+end_src
*** Olivetti mode
#+begin_src emacs-lisp
(setq olivetti-style nil ;; 'fancy
      olivetti-body-width 70)
(add-hook 'org-mode-hook (lambda () (olivetti-mode 1)))
#+end_src
Keybindings:
#+begin_src emacs-lisp
(map! :leader
      (:prefix "t"
               :desc "toggle Olivetti mode" "o" #'olivetti-mode
                ))
#+end_src

Disabling Olivetti Mode in ~Doom Docs Mode~:
#+begin_src emacs-lisp
(add-hook 'doom-docs-mode-hook (lambda () (olivetti-mode 'nil)))
#+end_src
*** Nov.el
Associate nov mode with epub files:
#+begin_src emacs-lisp
(add-to-list 'auto-mode-alist '("\\.epub\\'" . nov-mode))
#+end_src

Better fonts on nov.el:
#+begin_src emacs-lisp
(add-hook 'nov-mode-hook 'variable-pitch-mode)
(setq nov-text-width 80)
#+end_src
** Treemacs
#+begin_src emacs-lisp
(set-popup-rule! "^ ?\\*Treemacs" :ignore t)
(after! treemacs
  (define-key treemacs-mode-map [mouse-1] #'treemacs-single-click-expand-action)
  (treemacs-git-commit-diff-mode 't)
  (treemacs-git-mode 'extended)
  (treemacs-indent-guide-mode 't)
  (hide-mode-line-mode 't)
  (setq treemacs-indentation 1
        treemacs-indentation-string "┃"
        treemacs-width 25
        treemacs-wide-toggle-width 40
        treemacs-text-scale 1
        ))
#+end_src
** Vterm and Eshell
 * It seems like =vterm-shell= only requires the command of the shell, not the absolute path
#+begin_src emacs-lisp
(remove-hook 'vterm-mode-hook #'hide-mode-line-mode)
(after! vterm
(setq  vterm-shell "zsh"
       vterm-copy-exclude-prompt 't
       vterm-buffer-name-string "vterm %s"
       vterm-always-compile-module 't
       )
)
#+end_src
** Dired
*** Hide details by default
#+begin_src emacs-lisp
(add-hook 'dired-mode-hook
          (lambda ()
            (dired-hide-details-mode)))
#+end_src
*** Delete files to trash
(not strictly dired but most commonly used there)
#+begin_src emacs-lisp
(setq delete-by-moving-to-trash t
      trash-directory "~/.local/share/Trash/files/")
#+end_src
*** Keybinds
Evil mode motions and vim-like behavior
#+begin_src emacs-lisp
(evil-define-key 'normal dired-mode-map
        (kbd "h") 'dired-up-directory
        (kbd "l") 'dired-find-file
        (kbd "C") 'dired-do-copy
        (kbd "D") 'dired-do-delete
        (kbd "R") 'dired-do-rename
        (kbd "RET") #'dired-open-file
)
#+end_src

 * dired-open-file requires dired-open
#+begin_src emacs-lisp
(map! :leader
      (:prefix "d"
               :desc "toggle details"  "s" #'dired-hide-details-mode
               :desc "dired edit" "w" #'dired-toggle-read-only
               :desc "finish edit" "W" #'wdired-finish-edit
               :desc "cancel edit" "x" #'wdired-abort-changes
               :desc "open file" "o" #'dired-open-file
               ))
#+end_src
space+d+d is also defined below
*** Dired-single
Makes dired use a single buffer instead of new buffers for every directory
#+begin_src emacs-lisp
(defun my-dired-init ()
  "Bunch of stuff to run for dired, either immediately or when it's
   loaded."
  (define-key dired-mode-map [remap dired-find-file]
    'dired-single-buffer)
  (define-key dired-mode-map [remap dired-mouse-find-file-other-window]
    'dired-single-buffer-mouse)
  (define-key dired-mode-map [remap dired-up-directory]
    'dired-single-up-directory))

;; if dired's already loaded, then the keymap will be bound
(if (boundp 'dired-mode-map)
    ;; we're good to go; just add our bindings
    (my-dired-init)
  ;; it's not loaded yet, so add our bindings to the load-hook
  (add-hook 'dired-mode-hook 'my-dired-init))

(setq dired-single-use-magic-buffer t
      dired-single-magic-buffer-name "*dired*")

(map! :leader
      (:prefix "d"
               :desc "Open dired"  "d" (function
        (lambda nil (interactive)
        (dired-single-magic-buffer default-directory)))))

#+end_src
*** Dired subtree
Allows browsing subdirectories inside the same buffer
#+begin_src emacs-lisp
(use-package dired-subtree :ensure t)
(evil-define-key 'normal dired-mode-map
        (kbd "<tab>") 'dired-subtree-toggle
        (kbd "<backtab>") 'dired-subtree-cycle
)
(after! dired
  (setq dired-subtree-use-backgrounds nil))
#+end_src
*** Dired-open
#+begin_src emacs-lisp
(use-package dired-open
  :config
  (setq dired-open-extensions '(("png" . "ristretto")
                                ("docx" . "libreoffice")
                                ("docx" . "libreoffice")
                                ("ods" . "libreoffice")
                                ("odt" . "libreoffice")
                                ("xlsx" . "libreoffice")
                                ("pptx" . "libreoffice")
                                ("odp" . "libreoffice")
                                )))
#+end_src
*** Ranger mode
I currently have ranger disabled in init.el
#+begin_src emacs-lisp
(after! ranger
(setq ranger-show-hidden t
      ranger-max-parent-width 0.4
      ranger-width-preview 0.4
      ranger-max-preview-size 10
      ranger-dont-show-binary t
      ))
(add-hook 'ranger-mode-hook 'hide-mode-line-mode)
#+end_src
** Guix, scheme, and lisp hacking
*** Geiser REPL
#+begin_src emacs-lisp
(setq geiser-repl-query-on-kill-p nil)
;; (setq geiser-active-implementations '(guile))
;; (setq geiser-default-implementation '(guile))
#+end_src
** LSPs
#+begin_src emacs-lisp
(setq lsp-treemacs-errors-position-params `((side . right)))
#+end_src
*** Lua (fix for nixos)
#+begin_src emacs-lisp
  ;; ;;          doesn't work:
  ;; (setq
  ;;  lsp-clients-lua-language-server-bin (executable-find "lua-language-server")
  ;;       lsp-clients-emmy-lua-jar-path "~/.local.share/doom/lsp/emmy-lua/EmmyLua-LS-all.jar"
  ;;       lsp-clients-emmy-lua-java-path "~/.nix-profile/bin/java"
  ;;               )

;; ;;            doesn't work but different
;; (defun lslua-init ()
;;   "updates the lua lsp variable and runs lsp"
;;   (interactive)
;;   (setq lsp-clients-lua-language-server-bin (executable-find "lua-language-server"))
;;   (lsp)
;;   )
;; (add-hook 'lua-mode-hook #'lslua-init)

;; ;;           works?!
(after! lua-mode (setq lsp-clients-lua-language-server-bin (executable-find "lua-language-server"))
  (set-lsp-priority! 'lua-language-server 1))
#+end_src
**** TODO getting errors about tree sitter
**** TODO I still have to refresh lsp-mode and then lua-mode on first run
*** Python (also fix for nixos)
#+begin_src emacs-lisp
#+end_src
**** TODO getting errors about company mode
*** Nix LSP
#+begin_src emacs-lisp
(use-package lsp-mode
  :ensure t)

(use-package lsp-nix
  :ensure lsp-mode
  :after (lsp-mode)
  :demand t
  :custom
  (lsp-nix-nil-formatter ["nixpkgs-fmt"]))

(use-package nix-mode
  :hook (nix-mode . lsp-deferred)
  :ensure t)
#+end_src
* Packages
:PROPERTIES:
:header-args: :tangle packages.el
:END:
#+begin_src emacs-lisp
(package! tldr)
(package! pandoc)
(package! nov)
(package! evil-better-visual-line)
(package! catppuccin-theme)
(package! dired-subtree)
(package! dired-single)
(package! dired-open)
(package! mixed-pitch)
(package! olivetti)
(package! sicp)
(package! ob-racket :recipe
  (:host github
   :repo "DEADB17/ob-racket"
   :files ("*.el" "*.rkt")
   )
  )
#+end_src


home-manager/dotfiles/doom.d/config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(setq user-full-name "Alan Hettinger"
      user-mail-address "alan.hettinger@proton.me")

(load! "appearance")
(load! "writing/thesis")

(defun my/disable-scroll-bars (frame)
  (modify-frame-parameters frame
                           '((vertical-scroll-bars . nil)
                             (horizontal-scroll-bars . nil))))
(add-hook 'after-make-frame-functions 'my/disable-scroll-bars)

(use-package! evil-better-visual-line
  :config (evil-better-visual-line-on))
(+global-word-wrap-mode +1)

(setq split-height-threshold nil
      split-width-threshold 40)

(setq company-minimum-prefix-length 3)

(global-auto-revert-mode 1)
(setq global-auto-revert-non-file-buffers t)

(setq ispell-dictionary "en_US")
(setq +word-wrap-disabled-modes '(vterm-mode))

(setq evil-want-fine-undo t
      )

(map! :leader
      (:prefix "t"
       :desc "toggle modeline"  "m" #'hide-mode-line-mode)
      (:prefix "q"
       :desc "save and quit server-edit frame" "e" #'server-edit
       :desc "abort server-edit frame" "E" #'server-edit-abort)
      )

(map! :i "M-TAB" (cmds! (not (minibufferp)) #'company-complete-common))

(add-hook 'pdf-view-mode-hook
          (lambda () (pdf-view-auto-slice-minor-mode 1)))

(dolist (mode '(org-mode-hook
                markdown-mode-hook
                Info-mode-hook))
  (add-hook mode (lambda () (mixed-pitch-mode 1) ))
  (add-hook mode (lambda () (progn
                              (setq left-margin-width 4)
                              (setq right-margin-width 4)
                              (set-window-buffer nil (current-buffer))))))

(after! org
  (setq org-startup-folded t)
  (setq org-directory "~/Documents/Notes/org/")
  ;; makes info files linkable from org
  (add-to-list 'org-modules 'ol-info)

  (add-hook 'org-mode-hook
	    (setq org-pretty-entities t
		  org-hide-emphasis-markers t
		  org-adapt-indentation t))
  (setq org-ellipsis " ▼ ")
  )

(add-hook 'org-mode-hook
          (lambda () (setq-local company-idle-delay nil)))

(after! org
  (setq org-export-with-section-numbers nil
        org-export-with-toc nil
        ;; org-odt-preferred-output-format docx
        )
  )

(defun org-remove-headlines (backend)
  (org-map-entries (lambda () (delete-region (point-at-bol) (point-at-eol)))
                   "ignore"))

(add-hook 'org-export-before-processing-functions #'org-remove-headlines)


(org-babel-do-load-languages
 'org-babel-load-languages
 '((emacs-lisp . t)
   (lua . t)
   (scheme . t)
   (shell . t)
   (racket . t)
   ))

(after! org
  (require 'org-tempo)
  (add-to-list 'org-structure-template-alist '("sh" . "src shell"))
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("lua" . "src lua"))
  (add-to-list 'org-structure-template-alist '("sc" . "src scheme"))
  (add-to-list 'org-structure-template-alist '("r" . "src racket"))
  ;; automatically tangle certain config files on save:
  ;; (defun alan/org-babel-tangle-config ()
  ;;   (when (string-equal (buffer-file-name)
  ;;                       (expand-file-name "./config.org"))
  ;;     (let ((org-confirm-babel-evaluate nil))
  ;;       (org-babel-tangle))))
  ;; (add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook #'alan/org-babel-tangle-config)))
  )

(use-package! ob-racket
  :after org
  ;;   :config (add-hook 'ob-racket-pre-runtime-library-load-hook
  ;;               #'ob-racket-raco-make-runtime-library)
  )
;; (setq ob-racket-default-lang "sicp")

(after! org

  (setq org-latex-packages-alist '(("margin=2cm" "geometry" nil)))
  (setq org-cite-insert-processor 'citar
        org-cite-follow-processor 'citar
        org-cite-activate-processor 'citar
        citar-bibliography org-cite-global-bibliography
        org-cite-csl-styles-dir "~/Zotero/styles/"
        citar-citeproc-csl-styles-dir "~/Zotero/styles/"
        org-cite-export-processors
        '((latex . (csl "chicago-author-date.csl"))
          (odt . (csl "chicago-author-date.csl"))
          (t . (csl "chicago-author-date.csl")))
        )
  )

(after! writeroom-mode
  (setq +zen-text-scale 1)
  (setq writeroom-mode-line 't)
  )

(setq olivetti-style nil ;; 'fancy
      olivetti-body-width 70)
(add-hook 'org-mode-hook (lambda () (olivetti-mode 1)))

(map! :leader
      (:prefix "t"
       :desc "toggle Olivetti mode" "o" #'olivetti-mode
       ))

(add-hook 'doom-docs-mode-hook (lambda () (olivetti-mode 'nil)))

(add-to-list 'auto-mode-alist '("\\.epub\\'" . nov-mode))

(add-hook 'nov-mode-hook 'variable-pitch-mode)
(setq nov-text-width 80)

(set-popup-rule! "^ ?\\*Treemacs" :ignore t)
(after! treemacs
  (define-key treemacs-mode-map [mouse-1] #'treemacs-single-click-expand-action)
  (treemacs-git-commit-diff-mode 't)
  (treemacs-git-mode 'extended)
  (treemacs-indent-guide-mode 't)
  (hide-mode-line-mode 't)
  (setq treemacs-indentation 1
        treemacs-indentation-string "┃"
        treemacs-width 25
        treemacs-wide-toggle-width 40
        treemacs-text-scale 1
        ))

(remove-hook 'vterm-mode-hook #'hide-mode-line-mode)
(after! vterm
  (setq  vterm-shell "zsh"
         vterm-copy-exclude-prompt 't
         vterm-buffer-name-string "vterm %s"
         vterm-always-compile-module 't
         )
  )

(add-hook 'dired-mode-hook
          (lambda ()
            (dired-hide-details-mode)))

(setq delete-by-moving-to-trash t
      trash-directory "~/.local/share/Trash/files/")

(evil-define-key 'normal dired-mode-map
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-find-file
  (kbd "C") 'dired-do-copy
  (kbd "D") 'dired-do-delete
  (kbd "R") 'dired-do-rename
  (kbd "RET") #'dired-open-file
  )

(map! :leader
      (:prefix "d"
       :desc "toggle details"  "s" #'dired-hide-details-mode
       :desc "dired edit" "w" #'dired-toggle-read-only
       :desc "finish edit" "W" #'wdired-finish-edit
       :desc "cancel edit" "x" #'wdired-abort-changes
       :desc "open file" "o" #'dired-open-file
       ))

(defun my-dired-init ()
  "Bunch of stuff to run for dired, either immediately or when it's
   loaded."
  (define-key dired-mode-map [remap dired-find-file]
              'dired-single-buffer)
  (define-key dired-mode-map [remap dired-mouse-find-file-other-window]
              'dired-single-buffer-mouse)
  (define-key dired-mode-map [remap dired-up-directory]
              'dired-single-up-directory))

;; if dired's already loaded, then the keymap will be bound
(if (boundp 'dired-mode-map)
    ;; we're good to go; just add our bindings
    (my-dired-init)
  ;; it's not loaded yet, so add our bindings to the load-hook
  (add-hook 'dired-mode-hook 'my-dired-init))

(setq dired-single-use-magic-buffer t
      dired-single-magic-buffer-name "*dired*")

(map! :leader
      (:prefix "d"
       :desc "Open dired"  "d" (function
                                (lambda nil (interactive)
                                  (dired-single-magic-buffer default-directory)))))

(use-package dired-subtree :ensure t)
(evil-define-key 'normal dired-mode-map
  (kbd "<tab>") 'dired-subtree-toggle
  (kbd "<backtab>") 'dired-subtree-cycle
  )
(after! dired
  (setq dired-subtree-use-backgrounds nil))

(use-package dired-open
  :config
  (setq dired-open-extensions '(("png" . "ristretto")
                                ("docx" . "libreoffice")
                                ("docx" . "libreoffice")
                                ("ods" . "libreoffice")
                                ("odt" . "libreoffice")
                                ("xlsx" . "libreoffice")
                                ("pptx" . "libreoffice")
                                ("odp" . "libreoffice")
                                )))

(after! ranger
  (setq ranger-show-hidden t
        ranger-max-parent-width 0.4
        ranger-width-preview 0.4
        ranger-max-preview-size 10
        ranger-dont-show-binary t
        ))
(add-hook 'ranger-mode-hook 'hide-mode-line-mode)

(setq geiser-repl-query-on-kill-p nil)
;; (setq geiser-active-implementations '(guile))
;; (setq geiser-default-implementation '(guile))

(setq lsp-treemacs-errors-position-params `((side . right)))

(after! lua-mode (setq lsp-clients-lua-language-server-bin (executable-find "lua-language-server"))
  (set-lsp-priority! 'lua-language-server 1))



(use-package lsp-mode
  :ensure t)

(use-package lsp-nix
  :ensure lsp-mode
  :after (lsp-mode)
  :demand t
  :custom
  (lsp-nix-nil-formatter ["nixpkgs-fmt"]))

(use-package nix-mode
  :hook (nix-mode . lsp-deferred)
  :ensure t)
